### PR TITLE
fix: make port `0` bind to any available port

### LINKED ISSUE #2643: Support binding to port 0
If you ask linux to bind to port 0, it will automatically pick an open port for you. This can be really helpful when running multiple instances of ganache on the same machine. Currently, ganache says `Invalid port number '0'`

Please make ganache support binding to port 0.

It would also be helpful (though not necessary) if it wrote this port to a file somewhere.

### PR DESCRIPTION
**Pull request by davidmurdoch**
Specifying a port of `0` is a way to tell a server that you want the Operating System to let you bind to any open port. Ganache claimed to support binding to port 0, but would rudely yell at you if you tried: `Port should be >= 0 and < 65536. Received 0` :facepalm:! This PR rights our wrongs.

Using port `0` is useful when you don't care which port you bind to, you just want an open port.

You can use port `0` in a few ways:

### On the cli:

```console
$ ganache --port 0
…
RPC Listening on 127.0.0.1:35006
```

### In detach mode:

$$\color{#8b949e}\textsf{note: port 0 binding in detach mode doesn't yet provide a scripting friendly way of returning the port}$$
```console
$ ganache --detach --port 0
frozen_caramel_doughnut

$ ganache instances list
┌───────┬─────────────────────────┬──────────┬─────────┬─────────────────┬────────┐
│   PID │ Name                    │ Flavor   │ Version │ Host            │ Uptime │
├───────┼─────────────────────────┼──────────┼─────────┼─────────────────┼────────┤
│ 58078 │ frozen_caramel_doughnut │ ethereum │ 7.7.3   │ 127.0.0.1:35006 │    14s │
└───────┴─────────────────────────┴──────────┴─────────┴─────────────────┴────────┘
```

### and programmatically:

```javascript
import ganache from "ganache";

const server = ganache.server();
await server.listen(0);
console.log("Listening on port: ", server.address().port);
```

Fixes #2643 and #3805

### CODE REVIEW

File: .github/workflows/pr.yml [hunk 1]
```diff
@@ -47,7 +47,7 @@ jobs:
 
       - name: Check bundle size
         # this should match the os and version used in the release.yml
-        if: startsWith(matrix.os, 'ubuntu-20.04') && startsWith(matrix.node, '14.')
```
Comment (davidmurdoch) [inline .github/workflows/pr.yml]: This stops the `check bundle size` run from happening twice; we don't want it to run on `14.0.0`

File: src/packages/cli/src/cli.ts [hunk 2]
```diff
@@ -157,7 +158,7 @@ if (argv.action === "start") {
     // instance), so we need to notify that we are ready.
     const isDetachedInstance = process.send !== undefined;
     if (isDetachedInstance) {
-      notifyDetachedInstanceReady();
+      notifyDetachedInstanceReady(port);
```
Comment (davidmurdoch) [inline src/packages/cli/src/cli.ts:161]: < controversy >

File: src/packages/core/src/server.ts [hunk 3]
```diff
@@ -33,6 +34,13 @@ import WebsocketServer, { WebSocketCapableFlavor } from "./servers/ws-server";
 import HttpServer from "./servers/http-server";
 import Emittery from "emittery";
 
+// not using the "net" node package in order to avoid having to polyfill this
+// for the browser build.
+// isIPv4 taken from https://github.com/nodejs/node/blob/01323d50c4b24cf730a651d06ba20633905ecbed/lib/internal/net.js#L31
+const v4Seg = "(?:[0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])";
+const IPv4Reg = new RegExp(`^(${v4Seg}[.]){3}${v4Seg}$`);
+const isIPv4 = (s: string) => IPv4Reg.test(s);
```
Comment (davidmurdoch) [inline src/packages/core/src/server.ts:42]: While `server` isn't useful in the browser, our webpack step still processes it. I could have made it ignore the `net` package, or used a polyfill, but this code was so simple I didn't think it was necessary.

I think copy+paste is ultimately better than ignoring or partially polyfilling `net` because if we ever _do_ utilize the `net` package somewhere else the webpack process to succeed, but the actual probably won't work (because the polyfill is incomplete or just missing). Note: polyfilling the entire `net` package isn't possible, since we can't create actual sockets/pipes/etc in the browser.

You insights on this will be much appreciated!

File: src/packages/core/src/server.ts [hunk 4]
```diff
@@ -43,13 +51,13 @@ export type Callback = (err: Error | null) => void;
...
- *  * closing: `status === Status.closing` or `status & Status.closing !== 0`
- *  * closed: `status === Status.closed` or `status & Status.closed !== 0`
- *  * closing || closed: `status & Status.closingOrClosed !== 0` or `status & (Status.closing | Status.closed) !== 0`
+ *  * ready: `status === ServerStatus.ready` or `status & ServerStatus.ready !== 0`
+ *  * opening: `status === ServerStatus.opening` or `status & ServerStatus.opening !== 0`
+ *  * open: `status === ServerStatus.open` or `status & ServerStatus.open !== 0`
+ *  * opening || open: `status & ServerStatus.openingOrOpen !== 0` or `status & (ServerStatus.opening | ServerStatus.open) !== 0`
+ *  * closing: `status === ServerStatus.closing` or `status & ServerStatus.closing !== 0`
+ *  * closed: `status === ServerStatus.closed` or `status & ServerStatus.closed !== 0`
+ *  * closing || closed: `status & ServerStatus.closingOrClosed !== 0` or `status & (ServerStatus.closing | ServerStatus.closed) !== 0`
```
Comment (davidmurdoch) [inline src/packages/core/src/server.ts:60]: This is just a drive-by fix up. A version of this PR used `ServerStatus`, and I had tried to copy paste from the examples here... but it didn't work because the names were wrong.

Let me know if you want this in another PR.

File: src/packages/core/src/server.ts [hunk 5]
```diff
@@ -283,6 +292,19 @@ export class Server<
     }
   }
 
+  public address() {
```
Comment (davidmurdoch) [inline src/packages/core/src/server.ts:297]: This PR copies Node's own http `createServer`'s `address()` API: https://nodejs.org/api/net.html#net_server_address

With one exception: we do _not_ support named pipes and Unix domain sockets, so we do need to support returning a string.

File: src/packages/core/src/server.ts [hunk 6]
```diff
@@ -283,6 +292,19 @@ export class Server<
...
+    if (this.#listenSocket) {
```
Comment (davidmurdoch) [inline src/packages/core/src/server.ts:298]: My first iteration of this checked the server's `#status` to see if we were "Open". This `this.#listenSocket` check is effectively the same (the two properties are always set together). I chose this check instead only because passing `null` to `us_socket_local_port` triggers what seems to be undefined behavior in the uWS library, so this is slightly more explicit and less prone to bugs due to future refactoring.

File: src/packages/core/tests/server.test.ts [hunk 7]
```diff
@@ -369,8 +425,7 @@ describe("server", () => {
         "-0o1",
         "-0b1",
         "0o",
-        "0b",
-        0
+        "0b"
```
Comment (davidmurdoch) [inline src/packages/core/tests/server.test.ts:428]: Yes, these are both an encoding of `0` in octal and binary, respectively, but we don't support that, and neither does node's `http.createServer.listen(<port>)`

File: src/packages/cli/src/detach.ts [hunk 8]
```diff
@@ -113,11 +112,14 @@ export async function startDetachedInstance(
...
   child.stderr.pipe(process.stderr);
 
-  await new Promise<void>((resolve, reject) => {
-    child.on("message", message => {
-      if (message === READY_MESSAGE) {
-        resolve();
-      }
+  // Wait for the child process to send the "ready" message, which indicates
+  // that the Ganache server has started and is ready to receive RPC requests.
+  // It signals by sending the port number to which it was bound to back to us;
```
Comment (MicaiahReid) [inline src/packages/cli/src/detach.ts]:
```suggestion
  // It signals by sending the port number to which it was bound back to us;
```
I think?

File: src/packages/core/src/server.ts (hunk 3 above)
Comment (jeffsmale90) [reply src/packages/core/src/server.ts:42]: I think this is fine

File: src/packages/cli/src/detach.ts [hunk 9]
```diff
@@ -159,7 +161,7 @@ export async function startDetachedInstance(
     pid,
     name: createInstanceName(),
     host,
-    port,
```
Comment (MicaiahReid) [inline src/packages/cli/src/detach.ts]: Because we are no longer using the `port` variable, it can be deleted a few lines above

File: src/packages/core/src/server.ts (hunk 3 above)
Comment (MicaiahReid) [reply src/packages/core/src/server.ts:42]: I'm cool with the approach. I'm particularly cool with _not_ making our webpack process any more complicated

File: src/packages/cli/src/cli.ts (hunk 2 above)
Comment (jeffsmale90) [reply src/packages/cli/src/cli.ts:161]: I don't like this. Sending an object has _minimal_ additional overhead in terms of complexity and performance, and makes this statement much more obvious:

```
child.on("message", port => {
  resolve(port as number);
});
```

```
child.on("message", message => {
  if (message.message === READ_MESSAGE) {
    resolve(message.port as number);
  }
});
```

Granted, this area is not _super_ complicated, and the surrounding code is quite explanatory, but :shrug:./

Keen to hear other people's opinions...

**Review by jeffsmale90 — CHANGES_REQUESTED** [review]

File: src/packages/core/tests/server.test.ts [hunk 10]
```diff
@@ -336,15 +336,71 @@ describe("server", () => {
...
+          const address = s.address();
+          assert.notEqual(
+            address,
+            null,
+            "server.address() should not be null after listening"
+          );
+          assert.strictEqual(
+            address!.port,
+            +specificPort,
+            "server.address().port should be non-zero after server.listen(0)"
```
Comment (MicaiahReid) [inline src/packages/core/tests/server.test.ts]:
```suggestion
            "server.address().port should be equal to the specified port after server.listen(nonZeroPort)"
```
Or something like that, since this test isn't testing `server.listen(0)`

**Review by MicaiahReid — CHANGES_REQUESTED** [review]

File: src/packages/cli/src/cli.ts (hunk 2 above)
Comment (davidmurdoch) [reply src/packages/cli/src/cli.ts:161]: If we can't trust that `message` is what we expect it to be (since we have to check against `READY_MESSAGE`), then do we also need to check the entire payload? Basically:

```
try {
  if (message && typeof message === "object" && message.message === READ_MESSAGE && typeof message.port === "number" && isValidPort(message.port)) {
    resolve(message.port);
  }
} catch {
  // ignore unexpected messages
}
function isValidPort(port: number): boolean {
   // ...
}
```

File: .github/workflows/pr.yml [hunk 11]
```diff
@@ -43,7 +43,7 @@ jobs:
 
       - name: Check bundle size
         # this should match the os and version used in the release.yml
-        if: startsWith(matrix.os, 'ubuntu-22.04') && startsWith(matrix.node, '14.')
+        if: startsWith(matrix.os, 'ubuntu-22.04') && startsWith(matrix.node, '14.') && matrix.node != '14.0.0'
```
Comment (davidmurdoch) [inline .github/workflows/pr.yml:46]: This stops the check bundle size run from happening twice; we don't want it to run on 14.0.0

**Review by MicaiahReid — APPROVED** [review]
I think this is a fix because we claimed to support binding to port 0.

File: src/packages/cli/src/detach.ts [hunk 12]
```diff
@@ -113,11 +112,13 @@ export async function startDetachedInstance(
...
 
-  await new Promise<void>((resolve, reject) => {
-    child.on("message", message => {
-      if (message === READY_MESSAGE) {
-        resolve();
-      }
+  // Wait for the child process to send its port, which indicates that the
+  // Ganache server has started and is ready to receive RPC requests. It signals
+  // by sending the port number to which it was bound back to us; this is needed
+  // because Ganache may bind to a random port if the user specified port 0.
```
Comment (jeffsmale90) [inline src/packages/cli/src/detach.ts:118]: 👍

File: src/packages/cli/src/cli.ts (hunk 2 above)
Comment (jeffsmale90) [reply src/packages/cli/src/cli.ts:161]: After much discussions, I'm happy with the solution as is, with the updated comment in detach.ts

File: src/packages/core/tests/server.test.ts [hunk 13]
```diff
@@ -336,15 +336,71 @@ describe("server", () => {
...
+      );
+      await s.listen(zeroPort);
+
+      try {
+        const address = s.address();
+        assert.notEqual(
+          address,
+          null,
+          "server.address() should not be null after listening"
+        );
```
Comment (jeffsmale90) [inline src/packages/core/tests/server.test.ts:387]: Would it be worthwhile asserting that the value is a non-negative integer? Maybe not because we do ensure that sending a request there is successful 🤷

**Review by jeffsmale90 — APPROVED** [review]
